File: 数组array/easy-01-两数之和/twoSum.js
/**
* @param {number[]} nums
* @param {number} target
* @return {number[]}
*/
const twoSum = function(nums, target) {
    // 暴力枚举法 112 ms	41.5 MB	JavaScript
//    const len = nums.length;
//    for(let i=0; i< len; i++) {
//        for(let j=0; j<len,j!==i; j++) {
//            if((nums[i]+nums[j]) === target) return [i,j];
//        }
//    }
//    return [];

    // 方法二： map 64 ms	42.5 MB	JavaScript
    let map = new Map();
    for(let i = 0; i< nums.length;i++) {
        const remain = target - nums[i];
        if (map.has(remain)) {
            return [i,map.get(remain)];
        } else {
            map.set(nums[i],i);
        }
    }
    return [];

};

File: 数组array/easy-01-两数之和/twoSum.java
import java.util.HashMap;
import java.util.Map;

class Solution {

    public static int[] twoSum(int[] nums, int target) {
        // 方法一：暴力枚举  97 ms	41.6 MB	Java
//        int n = nums.length;
//        for ( int i = 0; i < n; ++i ) {
//            for ( int j = 0 ; j < n; ++j ) {
//                if ( j != i && nums[i] + nums [j] == target ) {
//                    return new int[]{ i, j};
//                }
//            }
//        }
//        return new int[0];

        // 方法二：哈希表 2 ms	41.8 MB	Java
        Map<Integer, Integer> resultMap = new HashMap<Integer, Integer>();
        for ( int i = 0; i < nums.length; ++i ) {
            if ( resultMap.containsKey(target - nums[i])) {
                return new int[]{resultMap.get(target - nums[i]),i};
            }
            resultMap.put(nums[i], i);
        }
        return new int[0];
    }


    public static void main(String[] args) {
        int[] nums = {2,7,11,15};
        int target = 9;
        int[] result = twoSum(nums, target);
        for ( int i = 0; i < result.length; ++i ) {
            System.out.println(result[i]);
        }
    }

}





